### PR TITLE
Bg 4892 transaction builder

### PR DESCRIPTION
**Pull request by argjv**

### CODE REVIEW

File: src/transaction_builder.js [hunk 1]
```diff
@@ -486,25 +488,27 @@ function TransactionBuilder (network, maximumFeeRate) {
...
-  this.bitcoinCash = false
-  this.bitcoinGold = false
-  this.tx = new Transaction()
+  this.tx = new Transaction(this.network.coin)
 }
 
 TransactionBuilder.prototype.enableBitcoinCash = function (enable) {
-  if (typeof enable === 'undefined') {
-    enable = true
+  if (typeof enable === 'undefined' || enable) {
```
Comment (arik-so) [inline src/transaction_builder.js]: how does this handle testnet bitcoin gold/cash?

File: src/coins.js [hunk 2]
```diff
@@ -1,28 +1,36 @@
...
   BTC: 'btc',
   BTG: 'btg',
+  LTC: 'ltc',
   ZEC: 'zec'
 }
 
-coins.isBitcoin = function (value) {
-  return typeforce.String(value) && value === coins.BTC
+coins.isBitcoin = function (network) {
+  return typeforce.String(network.coin) && network.coin === coins.BTC
```
Comment (johnoliverdriscoll) [inline src/coins.js]: `return typeforce.value(coins.BTC)(network.coin)`

File: src/coins.js [hunk 3]
```diff
@@ -1,28 +1,36 @@
...
+}
+
+coins.isBitcoinCash = function (network) {
+  return typeforce.String(network.coin) && network.coin === coins.BCH
```
Comment (johnoliverdriscoll) [inline src/coins.js]: `return typeforce.value(coins.BCH)(network.coin)`

File: src/coins.js [hunk 4]
```diff
@@ -1,28 +1,36 @@
...
+}
+
+coins.isBitcoinGold = function (network) {
+  return typeforce.String(network.coin) && network.coin === coins.BTG
```
Comment (johnoliverdriscoll) [inline src/coins.js]: `return typeforce.value(coins.BTG)(network.coin)`

File: src/coins.js [hunk 5]
```diff
@@ -1,28 +1,36 @@
...
 }
 
-coins.isBitcoinCash = function (value) {
-  return typeforce.String(value) && value === coins.BCH
+coins.isLitecoin = function (network) {
+  return typeforce.String(network.coin) && network.coin === coins.LTC
```
Comment (johnoliverdriscoll) [inline src/coins.js]: `return typeforce.value(coins.LTC)(network.coin)`

File: src/coins.js [hunk 6]
```diff
@@ -1,28 +1,36 @@
...
 }
 
-coins.isBitcoinGold = function (value) {
-  return typeforce.String(value) && value === coins.BTG
+coins.isZcash = function (network) {
+  return typeforce.String(network.coin) && network.coin === coins.ZEC
```
Comment (johnoliverdriscoll) [inline src/coins.js]: `return typeforce.value(coins.ZEC)(network.coin)`

File: src/coins.js [hunk 7]
```diff
@@ -1,28 +1,36 @@
...
 }
 
-coins.isZcash = function (value) {
-  return typeforce.String(value) && value === coins.ZEC
+coins.isValidCoin = function (network) {
+  return typeforce.String(network.coin) && network.coin in [coins.BTC, coins.BCH, coins.BTG, coins.LTC, coins.ZEC]
```
Comment (johnoliverdriscoll) [inline src/coins.js]: return typeforce.oneOf(coins.BTC, coins.BCH, coins.BTG, coins.LTC, coins.ZEC)(network.coin)

File: src/transaction_builder.js (hunk 1 above)
Comment (argjv) [reply src/transaction_builder.js]: I'm removing these methods since mutating the network in a transaction builder can lead to errors.

**Comment by argjv** [comment]
I've updated the pull request with the review comments and more unit tests for Zcash mainly. PTAL

File: src/coins.js [hunk 8]
```diff
@@ -10,27 +10,27 @@ const coins = {
...
 }
 
 coins.isZcash = function (network) {
-  return typeforce.String(network.coin) && network.coin === coins.ZEC
+  return typeforce.value(coins.ZEC)(network.coin)
 }
 
 coins.isValidCoin = function (network) {
-  return typeforce.String(network.coin) && network.coin in [coins.BTC, coins.BCH, coins.BTG, coins.LTC, coins.ZEC]
+  return typeforce.oneOf(coins.BTC, coins.BCH, coins.BTG, coins.LTC, coins.ZEC)(network.coin)
```
Comment (johnoliverdriscoll) [inline src/coins.js]: You actually need to convert each of the coins into a typeforce checker with typeforce.value(...)
```
return typeforce.oneOf(
  typeforce.value(coins.BTC),
  typeforce.value(coins.BCH),
  typeforce.value(coins.BTG),
  typeforce.value(coins.LTC),
  typeforce.value(coins.ZEC)
)(network.coin)
```
Sorry, this was my mistake that was carried over into this commit.

Another idea, in the spirit of code re-use, would be to incorporate typeforce.oneOf and our own coins.isXYZ helpers:
```
coins.isValidCoin = typeforce.oneOf(
  coins.isBitcoin,
  coins.isBitcoinCash,
  coins.isBitcoinGold,
  coins.isLitecoin,
  coins.isZcash
)
```

Comment (argjv) [reply src/coins.js]: Done

File: src/transaction_builder.js [hunk 9]
```diff
@@ -522,28 +513,78 @@ TransactionBuilder.prototype.setLockTime = function (locktime) {
...
   }
 
   // Copy transaction fields
-  txb.setVersion(transaction.version)
+  txb.setVersion(transaction.version, transaction.overwintered)
   txb.setLockTime(transaction.locktime)
 
+  if (coins.isZcash(txbNetwork)) {
+    // Copy Zcash overwinter fields. If the transaction builder is not for Zcash, they will be omitted
+    txb.setVersionGroupId(transaction.versionGroupId)
```
Comment (arik-so) [inline src/transaction_builder.js]: please add zcash version checks and only call the methods based on those

**Pull request closed without merging**